yesses/utils.py

```diff
@@ -21,5 +27,138 @@ def patched_create_connection(address, *args, **kwargs):
+    """
+    filtered_origins = dict()
+    for origin in origins:
+        parsed_url = UrlParser(origin['url'])
+        with force_ip_connection(origin['domain'], origin['ip']):
+            r = requests.get(parsed_url.url_without_path)
+            forwareded_parsed_url = UrlParser(r.url)
+            if forwareded_parsed_url.url_without_path not in filtered_origins.keys():
+                url = f"{forwareded_parsed_url.url_without_path}/"
+                filtered_origins[forwareded_parsed_url.url_without_path] = {'url': url,
+                                                                            'domain': forwareded_parsed_url.domain,
+                                                                            'ip': origin['ip']}
+    return filtered_origins
+
+
+def request_is_text(r: requests.Response) -> bool:
+    if 'content-type' not in r.headers:
+        return False
+    if re.search(r"(^text/.*|^image/svg\+xml$)", r.headers['content-type']):
+        return True
+    return False
+
+
+def read_file(list: str) -> List[str]:
+    with open(list) as file:
+        dir_list = file.readlines()
+        dir_list = [line.strip('\n') for line in dir_list if not line.startswith('#')]
+    return dir_list
+
+
+def convert_header(r: requests.Response) -> List[str]:
+    response = []
+    for key, value in r.headers.items():
+        response.append(f"{key}: {value}")
+    return response
+
+
+class UrlParser:
+    STANDARD_PORTS = {'http': 80, 'https': 443}
+
+    def __init__(self, url: str):
+        self.parsed = urlparse(url)  # type: ParseResult
+        self.netloc = self.parsed.netloc  # type: str
+        self.path = self.parsed.path  # type: str
+        self.arguments = self.parsed.query  # type: str
+        self.scheme = self.parsed.scheme  # type: str
+
+        tmp_path = self.path
+        if not tmp_path.startswith('/'):
+            tmp_path = f"/{tmp_path}"
+
+        # calculate the depth of the path
+        self.path_depth = len(tmp_path.split('/')) - 1
+        if tmp_path.endswith('/'):
+            self.path_depth -= 1
+
+        # concatenate the path with the GET parameters
+        self.path_with_args = tmp_path  # type: str
+        if self.parsed.query != '':
+            self.path_with_args = f"{tmp_path}?{self.arguments}"
+
+        # cut the file ending from the path
+        index = self.parsed.path.rfind('.')
+        if index != -1:
+            self.file_ending = self.parsed.path[index:]
+        else:
+            self.file_ending = ''
+
+        # retrieve the base domain (domain which was passed without the port and 'www.' prefix)
+        tmp = self.parsed.netloc.split(':')
+        self.base_domain = tmp[0]
+        split = self.base_domain.split('.')
+        if split[0] == "www":
+            self.base_domain = '.'.join(split[1:])
+
+        # the url with the protocol and port (if no port is specified use a standard port)
+        self.url_without_path = self.parsed.netloc
```

Comment:
Wouldn't that be the origin?

Reply:
I have just googled it and according to this [website](https://developer.mozilla.org/en-US/docs/Web/API/URL/origin), you are right. I will rename it.